desktop/notification/notify.go

```diff
@@ -0,0 +1,142 @@
+// re-started and should be used for as long as the sender process is alive, as
+// sending the identifier across session startup boundary has no chance of
+// working correctly.
+//
+// A notification may optionally carry a number of hints that further customize it
+// in a specific way. Refer to various hint constructors for details.
+//
+// A notification may optionally also carry one of several actions. If
+// supported, actions can be invoked by the user, broadcasting a notification
+// response back to the session. This mechanism only works if there is someone
+// listening for the action being triggered.
+//
+// In all cases, the specific notification must take into account the
+// capabilities of the server. For instance, if a server does not support body
+// markup, then such markup is not automatically stripped by either the client
+// or the server.
+type Message struct {
+	AppName       string
+	Icon          string
+	Summary       string
+	Body          string
+	ExpireTimeout int32 // XXX: consider using time.Duration instead?
+	ReplacesID    ID
+	Actions       []Action
+	Hints         []Hint
+}
+
+// ID is the opaque identifier of a notification assigned by the server.
+//
+// Notifications with known identifiers can be closed or updated. The identifier
+// is valid within one desktop session and should not be used unless the calling
+// process initially sent the message.
+type ID uint32
+
+// Action describes a single notification action.
+//
+// ActionKey is returned in an D-Bus signal when an action is activated by the
+// user. The text must be localized for the appropriate language.
+type Action struct {
+	ActionKey     string
+	LocalizedText string
+}
+
+// Hint describes supplementeary information that may be used by the server.
+//
+// Various helpers create hint objects of specifc purpose.
+type Hint struct {
+	Name  string
+	Value interface{}
+}
+
+// ServerCapability describes a single capability of the notification server.
+type ServerCapability string
+
+// CloseReason indicates why a notification message was closed.
+type CloseReason uint32
+
+const (
+	// CloseReasonExpired indicates that a notification message has expired.
+	CloseReasonExpired CloseReason = 1
+	// CloseReasonDismissed indicates that a notification message was dismissed by the user.
+	CloseReasonDismissed CloseReason = 2
+	// CloseReasonClosed indicates that a notification message was closed with an API call.
+	CloseReasonClosed CloseReason = 3
+	// CloseReasonUndefined indicates that no other well-known reason applies.
+	CloseReasonUndefined CloseReason = 4
+)
+
+// String implements the Stringer interface.
+func (r CloseReason) String() string {
+	switch r {
+	case CloseReasonExpired:
+		return "expired"
+	case CloseReasonDismissed:
+		return "dismissed"
+	case CloseReasonClosed:
+		return "closed"
+	case CloseReasonUndefined:
+		return "undefined"
+	default:
+		return fmt.Sprintf("CloseReason(%d)", uint32(r))
```

Comment:
Same question about the default case and its handling server-side.

Reply:
Same response, it's just a stringer, it's never sent to the server.